src/ucp/core/ucp_mm.c

```diff
@@ -693,6 +694,70 @@ void ucp_mpool_obj_init(ucs_mpool_t *mp, void *obj, void *chunk)
     elem_hdr->memh = chunk_hdr->memh;
 }
 
+static inline ucs_status_t
+ucp_rndv_frag_malloc_mpools(ucs_mpool_t *mp, size_t *size_p, void **chunk_p)
+{
+    ucp_rndv_mpool_priv_t *mpriv = ucs_mpool_priv(mp);
+    ucp_mem_desc_t *chunk_hdr;
+    ucp_mem_h memh;
+    ucs_status_t status;
+    ucp_mem_map_params_t mem_params;
+
+    /* metadata */
+    chunk_hdr = ucs_malloc((*size_p + sizeof(*chunk_hdr)), "chunk_hdr");
+    if (chunk_hdr == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    /* payload */
+    /* Need to get default flags from ucp_mem_map_params2uct_flags() */
+    mem_params.field_mask = 0;
+    status = ucp_mem_map_common(mpriv->worker->context, NULL,
+                                (mpriv->frag_size * mpriv->num_frags),
+                                mpriv->mem_type,
+                                ucp_mem_map_params2uct_flags(&mem_params),
+                                1, ucs_mpool_name(mp), &memh);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    chunk_hdr->memh  = memh;
+    chunk_hdr->index = 0; /* without index, index derivation in obj init
+                             requires deriving chunk padding */
+    *chunk_p         = chunk_hdr + 1;
+
+out:
+    return status;
+}
+
+static inline void
+ucp_rndv_frag_free_mpools(ucs_mpool_t *mp, void *chunk)
+{
+    ucp_mem_desc_t *chunk_hdr;
+    ucp_worker_h worker;
+    ucp_rndv_mpool_priv_t *mpriv;
+
+    mpriv  = (ucp_rndv_mpool_priv_t *)ucs_mpool_priv(mp);
+    worker = mpriv->worker;
+
+    chunk_hdr = (ucp_mem_desc_t*)chunk - 1;
+    ucp_mem_unmap_common(worker->context, chunk_hdr->memh);
+    ucs_free(chunk_hdr);
+}
+
+void ucp_frag_mpool_obj_init(ucs_mpool_t *mp, void *obj, void *chunk)
+{
+    ucp_mem_desc_t *elem_hdr     = obj;
+    ucp_mem_desc_t *chunk_hdr    = ((ucp_mem_desc_t*)chunk - 1);
+    ucp_rndv_mpool_priv_t *mpriv = ucs_mpool_priv(mp);
+
+    elem_hdr->memh   = chunk_hdr->memh;
+    elem_hdr->ptr    = (void*)((char*)chunk_hdr->memh->address
+                               + (chunk_hdr->index * mpriv->frag_size));
```

Comment:
use UCS_PTR_BYTE_OFFSET to simplify
can also use helper var "void *frag_base = chunk_hdr->memh->address`